specs/hierarchical_consensus.md

```diff
@@ -0,0 +1,1082 @@
+    // supply is left for the subnet (i.e. balance = 0).
+    // (methodNum = 5)
+    Kill()
+    
+    // Kill expects as source the address of the subnet actor for which the
+    // checkpoint is being committed. The function performs some basic checks
+    // to ensure that checkpoint is valid and it persist it in the SCA state.
+    // (methodNum = 6)
+    CommitChildCheckpoint(ch Checkpoint)
+    
+    // Fund can be called by any user in a subnet and it injects 
+    // the `value` of native tokens included in the message to the source's address in
+    // the child subnet given as argument.
+    // (methodNum = 7)
+    Fund(address.SubnetID)
+    
+    // Release can be called by any user in a subnet to release the amount 
+    // of native tokens included in `value` from its own address in the
+    // subnet to the address in the parent. 
+    // (methodNum = 8)
+    Release()
+    
+    // SendCross can be called by any user in the subnet to send 
+    // an arbitrary cross-message to any other subnet in
+    // the hierarchy.
+    // (methodNum = 9)
+    SendCross(msg CrossMsgParams)
+    
+    // ApplyMessage can only be called as an `ImplicitMessage` by
+    // the `SystemActor` and is used to perform the execution of a 
+    // of cross-net messages in the subnet. 
+    // (methodNum = 10)
+    ApplyMessage(msg CrossMsgParams)
+    
+    // InitAtomicExec can be called by users to initiate an
+    // atomic execution with some other subnet.
+    // (methodNum = 11)
+    InitAtomicExec(params AtomicExecParams)
+    
+    // SubmitAtomicExec has to be called by all participants in an
+    // atomic execution to submit their results and trigger the 
+    // propagation of the output (or the abortion) of the execution
+    // to the corresponding subnets.
+    // (methodNum = 12)
+    SubmitAtomicExec(submit SubmitExecParams)
+    
+    // WIP: Not implemented yet.
+    // (methodNum = 13)
+    ReportMisbehavior()
+    
+    // WIP: Not implemented yet.
+    // (methodNum = 14)
+    Save()
+```
+
+> All function parameters from the above snippet of code are fully specified in the detailed description of the different protocols.
+
+##### SCA State
+> State of the SCA
+```go
+type SCAState struct {
+    // ID of the current network
+    NetworkName address.SubnetID
+    // Total of active subnets spawwned from this one
+    TotalSubnets uint64
+    // Minimum stake required to create a new subnet
+    CollateralThreshold abi.TokenAmount
+    // List of subnets
+    Subnets cid.Cid // HAMT[cid.Cid]Subnet
+    // Checkpoint period in number of epochs for the subnet
+    CheckPeriod abi.ChainEpoch
+    // Checkpoints committed in SCA
+    Checkpoints cid.Cid // HAMT[epoch]Checkpoint
+    // CheckMsgMetaRegistry
+    // Stores information about the list of messages and child msgMetas being
+    // propagated in checkpoints to the top of the hierarchy.
+    CheckMsgsRegistry cid.Cid // HAMT[cid]CrossMsgs
```

Comment:
Just looking at this state and the SCA interface, it is not clear to me when this registry gets cleared of messages that have been applied, if ever. I suppose when `AppliedBottomUpNonce` gets adjusted, which is hopefully explained under checkpointing (I have to read that again) but just saying that the interface doesn't reflect this to me.

Reply:
Oh I see from the code that it's `ApplyMsg`. And that some kind of replacement is happening in the registry.